IdentityCore/src/cache/key/MSIDMacKeychainAccountCacheKey.m

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "MSIDMacKeychainAccountCacheKey.h"
+
+static NSString *keyDelimiter = @"-";
+static NSInteger kAccountTypePrefix = 1000;
+
+@implementation MSIDMacKeychainAccountCacheKey
+
+- (instancetype)initWithHomeAccountId:(NSString *)homeAccountId
+                          environment:(NSString *)environment
+                                realm:(NSString *)realm
+                                 type:(nullable NSNumber *)type
+{
+    self = [super init];
+    
+    if (self)
+    {
+        _homeAccountId = homeAccountId;
+        _environment = environment;
+        _realm = realm;
+        _accountType = type;
+    }
+    
+    return self;
+}
+
+- (NSData *)generic
+{
+    return [self.username.msidTrimmedString.lowercaseString dataUsingEncoding:NSUTF8StringEncoding];
+}
+
+- (NSNumber *)type
+{
+    if (_accountType)
+    {
+        return @([_accountType integerValue] + kAccountTypePrefix);
+    }
+    else
+    {
+        return nil;
+    }
+}
+
+- (NSString *)account
+{
+    if (self.homeAccountId.length && self.environment.length)
```

Comment:
Please reuse existing base implementation when possible :) (MSIDDefaultAccountCacheKey)

Reply:
updated

Reply:
Why can't we reuse existing implementation from parent class, it seems to do the same:

```
- (NSString *)account
{
    NSString *uniqueId = self.homeAccountId.msidTrimmedString.lowercaseString;

    return [NSString stringWithFormat:@"%@%@%@",
            uniqueId, keyDelimiter, self.environment];
}
```

Reply:
Similar to the type method update, I elimiated the subclass and added an `if()` for the account method. A `readAllAccounts:` would cause an account attribute string of `@"-"` and fail to match any accounts.

Reply:
If you don't know all attributes that compose the primary key (e.g. kSecAttrAccount or kSecAttrService), it shouldn't be using the key. There's a query subclass that does exactly that that you should be using instead.

Reply:
This change to the key object was eliminated, and the query object is now being used instead of the key.